File: src/batch_process/get_similarity_pairs.py
import pyspark
from pyspark.sql.window import Window
from pyspark.sql.types import *
from pyspark.sql.functions import udf, struct, col, rank
from pyspark.sql import SQLContext
from pyspark import SparkContext
from pyspark.sql.session import SparkSession

sc = SparkContext.getOrCreate()
spark = SparkSession(sc)

df = spark.read.text("/Users/adawang/Desktop/sample-s2-records")

def get_id(line):
    parenthesis = "\""
    paper_id_tag = "\"id\""
    id_label_start = line.find(paper_id_tag)
    id_tag_start = id_label_start + 6
    id_tag_end = line.find(parenthesis, id_tag_start)
    id_tag = line[id_tag_start -1 :id_tag_end + 1] 
    return id_tag

def get_title(line):
    # look for the title of the paper and return the tag 
    paper_title_tag = "\"title\""
    parenthesis = "\""
    
    title_label_start = line.find(paper_title_tag) # index for the title label start 
    title_tag_start = title_label_start + 9
    title_tag_end = line.find(parenthesis+",\"", title_tag_start) 
    title_tag = line[title_tag_start:title_tag_end]
    return title_tag

def adding_ids(df):
    add_ids = df.withColumn("id", get_id_udf(df.value))
    return add_ids

def get_title(line):
    '''
    This function takes the raw data dataframe and get the title column for the data
    ''' 
    paper_title_tag = "\"title\""
    parenthesis = "\""
    
    title_label_start = line.find(paper_title_tag) 
    title_tag_start = title_label_start + 9
    title_tag_end = line.find(parenthesis+",\"", title_tag_start) 
    title_tag = line[title_tag_start:title_tag_end]
    
    return title_tag

def adding_ids(df):
    '''
    This function takes the raw data dataframe and adds on an id column for the data
    '''
    add_ids = df.withColumn("id", get_id_udf(df.value))
    return add_ids

def adding_titles(df):
    '''
    This function takes the raw data dataframe and adds on an title column for the data
    '''
    add_titles = df.withColumn("title", get_title_udf(df.value))
    return add_titles


def drop_values(df):
    '''
    This function takes the dataframe and drops the value column
    '''
    return df.drop(df.value)

get_id_udf = udf(lambda line: get_id(line), StringType())
get_title_udf = udf(lambda line: get_title(line), StringType())

df = adding_ids(df)
df = adding_titles(df)
df = drop_values(df)

df.show()
df.cache()

from pyspark.ml import Pipeline
from pyspark.ml.feature import RegexTokenizer, NGram, HashingTF, MinHashLSH
import pyspark.sql.functions as f

model = Pipeline(stages = [RegexTokenizer(pattern = "", inputCol = "title", outputCol = "tokens", minTokenLength = 1), 
                           NGram(n = 3, inputCol = "tokens", outputCol = "ngrams"),HashingTF(inputCol = "ngrams", outputCol = "vectors"), MinHashLSH(inputCol = "vectors", outputCol = "lsh", numHashTables = 10)]).fit(df)

df_hashed = model.transform(df)


df_matches = model.stages[-1].approxSimilarityJoin(df_hashed, df_hashed, 0.9)

#show all matches (including duplicates)
df_matches.select(f.col('datasetA.id').alias('id_A'),
                 f.col('datasetB.id').alias('id_B'),
                 f.col('distCol')).show()

#show non-duplicate matches
df_matches.select(f.col('datasetA.id').alias('id_A'),
                 f.col('datasetB.id').alias('id_B'),
                 f.col('distCol')).filter('id_A < id_B').show()

File: src/batch_process/get_similarity_pairs.scala
import org.apache.spark.sql.types._
import org.apache.spark.ml.feature._
import org.apache.spark.ml.linalg._

val df = spark.read.option("delimiter","\t").csv("/Users/adawang/Desktop/sample-s2-records")
val dfUsed = df.select(col("_c1").as("id"), col("_c2").as("title")).filter(col("title") !== null)
dfUsed.show()

// Tokenize
val tokenizer = new Tokenizer().setInputCol("title").setOutputCol("tokens")

// 3gram
val ngram = new NGram().setN(3).setInputCol("tokens").setOutputCol("ngrams")

// vectorization
val vectorizer = new CountVectorizer().setInputCol("ngrams").setOutputCol("vectors")

// final pipeline
val pipelineTV = new Pipeline().setStages(Array(tokenizer, ngram, vectorizer))

val modelTV = pipelineTV.fit(dfused)

val isNoneZeroVector = udf({v: Vector => v.numNonzeros > 0}, DataTypes.BooleanType)

val df_TV = modelTV.transform(dfUsed).filter(isNoneZeroVector(col("features")))

val lsh = new MinHashLSH().setNumHashTables(10).setInputCol("features").setOutputCol("hashValues")
val pipelineLSH = new Pipeline().setStages(Array(lsh))
val modelLSH = pipelineLSH.fit(df_TV)

val df_hashed = modelLSH.transform(df_TV)

val df_matches = modelLSH.stages.last.asInstanceOf[MinHashLSHModel].approxSimilarityJoin(df_hashed, df_hashed, 0.9)
df_matches.show()
